### PR TITLE
fix(zod): reject non-object Zod v3 strict schema roots

### PR DESCRIPTION
**Pull request by HAYDEN-OAI**
## Summary

- Reject Zod v3 Structured Outputs schemas whose parsed root is not `type: "object"`, using the same synchronous error as the existing Zod v4 path across `zodResponseFormat`, `zodTextFormat`, `zodFunction`, and `zodResponsesFunction`.
- Gate validation exclusively on `openaiStrictMode` and perform it before root naming, schema-definition expansion, or `$ref` wrapping, preserving valid named, lazy, recursive, refined, transformed, and defaulted object roots.
- Add a focused 77-case regression matrix covering primitive/array/object-union/discriminated-union/nullable/optional/intersection/any roots, Zod v3/v4 behavior, nested unions and arrays, escaped schema definitions, every generic converter target, and unchanged Realtime behavior.

## Why

The vendored Zod v3 converter previously returned invalid strict root schemas unchecked. The four strict public helpers therefore emitted schemas that Structured Outputs cannot accept, while the equivalent Zod v4 conversion already failed synchronously. Valid Zod v4 optional object roots remain accepted because v4 normalizes them to an actual root object; the unsupported Zod v3 representation is correctly rejected.

## Validation

- Regression-first: the new focused suite produced **36 failures** against unchanged `main` before the fix.
- `pnpm test -- tests/helpers/zod-root-schema-validation.test.ts` — **77 passing tests**.
- `pnpm test:unit` — **2,167 passing tests across 71 files**.
- `pnpm format` and `pnpm lint`.
- `pnpm exec tsc`.
- `pnpm build`, including CommonJS/ESM SDK and Bedrock import checks.
- `./node_modules/typescript-4-9/bin/tsc --project dist/src/tsconfig.json --noEmit`.
- `./node_modules/typescript/bin/tsc --project dist/src/tsconfig.json --noEmit`.
- `pnpm exec publint dist`.
- CI-equivalent ATTW package analysis and `node scripts/utils/attw-report.cjs` — `Types ok!`.
- `node --experimental-strip-types scripts/test-packed-package.ts` — offline CommonJS/ESM consumer and both compiler versions; **263/300 source checks across 1,200 source maps**.
- Built-package smoke check confirms all four strict helpers reject primitive roots while `zodRealtimeFunction` remains non-strict.
- Both changed files are handwritten, have no generated marker, and are outside the generated/legacy allowlist.

### CODE REVIEW

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
Reviewed exact head `e1f2b1d97c6da364550aba75c8c6c76658c235ba`. The documented strict helpers are correct, but two Low internal strict-mode consistency gaps remain:

1. **Validate the effective returned root after name wrapping.** [`validateStrictRootSchema(main)`](https://github.com/openai/openai-node/blob/e1f2b1d97c6da364550aba75c8c6c76658c235ba/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L42-L54) accepts an object, after which the default named `nameStrategy: 'ref'` path [returns a root `$ref`](https://github.com/openai/openai-node/blob/e1f2b1d97c6da364550aba75c8c6c76658c235ba/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L100-L130), which strict Structured Outputs forbids. Public helpers use `duplicate-ref`, so this is limited to the package-importable `_vendor` converter and its internal `openaiStrictMode` option. Validate the final structural root or reject this strict name strategy; add direct strict named-object coverage.

2. **Reject nullable OpenAPI roots in the strict predicate.** The new [type-only check](https://github.com/openai/openai-node/blob/e1f2b1d97c6da364550aba75c8c6c76658c235ba/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L8-L16) accepts the OpenAPI3 nullable parser's `{ type: 'object', nullable: true }`, even though the root admits `null`. Public strict helpers do not select OpenAPI3, so this has the same narrow direct-converter scope. Make the predicate semantic enough to reject `nullable: true` and add a direct strict OpenAPI nullable-object regression.

No supported-helper/API issue was found; the focused matrix and strict thermo are clean, and exact CI is fully green.

**Review by jbeckwith-oai — APPROVED** [review]
Re-reviewed the full merge-base diff at exact head d2d73fe2eeb52244d2204410a755182dc2f2c0db. No findings. The strict-root guard now validates both the parsed and final wrapped schema, rejects OpenAPI nullable object roots, preserves supported strict-helper and non-strict behavior, and has focused parity coverage. Strict thermo and exact CI are clean.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 1]
```diff
@@ -40,6 +55,8 @@ const zodToJsonSchema = <Target extends Targets = 'jsonSchema7'>(
       false,
     ) ?? {};
 
+  validateStrictRootSchema(main, refs.openaiStrictMode);
```
[the remark this answers is not in the document]
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed on current head `28947e35411831daacced294f89c9ddaa29db174`. Strict conversion now materializes the supplied definition map before validating the root, follows only bounded same-document own-property definition references, and inlines the concrete object root without mutating caller-owned or frozen definitions. Regressions cover both directly registered and lazy registered Zod v3 object roots, recursive objects, alias chains, escaped names, all four strict public helpers, and Zod v4 parity; external/missing/cyclic/accessor-backed references, nullable or non-object roots, and unsafe metadata remain rejected.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 2]
```diff
@@ -5,6 +5,120 @@ import { parseDef } from './parseDef';
...
+  if (typeof rootReference?.value !== 'string') {
+    return schema;
+  }
+
+  const resolved = resolveStrictRootSchemaReference(schema, definitions, `${definitionPath.join('/')}/`);
+  validateStrictRootSchema(resolved, openaiStrictMode);
+  validateRootSchemaDataProperties(schema);
+  validateRootSchemaDataProperties(resolved);
+
+  const rootMetadata = { ...schema } as Record<string, unknown>;
```
Comment (jbeckwith-oai) [inline src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: [Medium] This shallow merge activates every enumerable sibling of the original root `$ref`. Under the default draft-07 target, assertion siblings were ignored; after materialization, `properties`, `required`, or `additionalProperties` become active and can replace the resolved strict object's values (`additionalProperties: true` can overwrite `false`, and sibling `properties` can discard referenced properties). Preserve only an explicit annotation allowlist when removing the top-level ref, or reject/semantically compose assertion siblings for targets where they apply. Add conflicting `properties` and `additionalProperties` regressions proving the referenced object remains authoritative.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 3]
```diff
@@ -5,6 +5,120 @@ import { parseDef } from './parseDef';
...
+    throw new Error(
+      `Root schema must have type: 'object' but got type: ${actualType ? `'${actualType}'` : 'undefined'}`,
+    );
+  }
+  if (Object.getOwnPropertyDescriptor(schema, '$ref')) {
+    throw new Error("Root schema must be a concrete object and cannot contain '$ref'");
+  }
+}
+
+function validateRootSchemaDataProperties(schema: object): void {
```
Comment (jbeckwith-oai) [inline src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: [Medium, strict thermo] This adds a second JSON-Schema reference engine—descriptor policy, alias walking, cycle/path handling, and materialization—beside the canonical `Refs.seen` / `parseDef(..., forceResolution)` machinery. The converter grows 128→251 lines for a root-type guard, with a 599-line test suite dominated by this new resolver, and the sibling-semantics bug below demonstrates the ownership risk. Resolve the strict root at the typed Zod-definition boundary via a root/force-resolution mode in `parseDef`; if a forced override still returns `$ref`, fail synchronously instead of following output reference strings. Keep focused public registered/lazy/recursive/naming/non-strict regressions and remove the custom resolver's implementation-detail matrix.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
Found one Medium reference-semantics bug and one Medium strict-thermo/architecture blocker in the exact-head diff. The earlier named-ref and nullable-root issues remain fixed; exact CI is green.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts (hunk 2 above)
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in `0f21e1aeb524385a343a36a87bbf5c1393ca6cf8`. Strict roots now enter the canonical forced `parseDef` path as concrete Zod objects, so root `$ref` siblings are never copied or shallow-merged. Separate regressions prove conflicting `properties` cannot replace the referenced object’s properties and `additionalProperties: true` cannot override `false`; a forced override that still returns a root `$ref` is rejected synchronously. Verified: **91 focused root-schema tests**, **232 helper tests each on Node 22.23.2 and 24.19.0**, and **2,181 full handwritten tests**.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts (hunk 3 above)
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in `0f21e1aeb524385a343a36a87bbf5c1393ca6cf8`. Removed the custom descriptor/JSON-pointer/alias/cycle/materialization resolver entirely. Strict conversion now uses the existing `parseDef(..., refs.openaiStrictMode === true)` / `Refs.seen` machinery, including canonical lazy traversal, valid recursive child references, and synchronous rejection if a forced override still returns `$ref`; non-strict parsing remains unchanged. The strict-only definitions clone preserves frozen caller definitions. Converter: **251 → 157 lines**; focused public root suite: **599 → 299 lines**, retaining direct/lazy/recursive/registered/frozen/naming/non-strict/Zod-v4 coverage. Verified: **91 focused tests**, **2,181 handwritten tests**, **232 helper tests plus packed-package checks on both Node 22.23.2 and 24.19.0**, lint/build, and TypeScript 4.9/6.0.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 4]
```diff
@@ -5,6 +5,30 @@ import { parseDef } from './parseDef';
 import { getRefs } from './Refs';
 import { zodDef, isEmptyObj } from './util';
 
+function validateStrictRootSchema(schema: object, openaiStrictMode: boolean | undefined): void {
+  if (!openaiStrictMode) {
+    return;
+  }
+
+  const type = 'type' in schema ? schema.type : undefined;
```
Comment (jbeckwith-oai) [inline src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: [Medium] These `in` checks and ordinary reads validate prototype-visible/non-enumerable properties, but JSON/request serialization emits only own enumerable properties. An override can return inherited or non-enumerable `type: 'object'`, pass both validations, then serialize without `type`; conversely invisible inherited/non-enumerable `$ref` or `nullable` can reject even though those keywords would not be sent. This path is package-published and the exported override contract is explicitly covered by this PR. Validate a plain serialization-equivalent root or inspect own enumerable data descriptors (and reject accessor-backed validation keys if desired), then add inherited/non-enumerable `type`, `$ref`, and `nullable` regressions that assert the emitted schema.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
Found one Medium strict-root serialization correctness issue in the exact-head diff. The prior reference-materialization problems are fixed and strict thermo passes; the sole live ecosystem failure is an unrelated external service-auth 500.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts (hunk 4 above)
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in 46eb7df096dffec08dc706bb9522a6caa108eebf. Strict-root validation now snapshots own property descriptors, considers only enumerable data properties for `type`, `$ref`, and `nullable`, and rejects enumerable accessor-backed validation keywords without invoking getters. Added six parameterized inherited/non-enumerable regressions that assert actual `JSON.stringify` output, plus three getter-side-effect regressions; all nine failed against the previous head and now pass. Verified: 100 focused tests; all 2,190 handwritten tests across 71 files; lint/format; current TypeScript; distributable build; published-source TypeScript 4.9 and 6.0 checks; `publint`; and the packed CommonJS/ESM consumer check (263 source checks across 1,200 source maps). Existing plain/named/lazy/recursive/frozen-definition/OpenAPI/non-strict/four-helper coverage remains green; only the two existing handwritten files changed.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
Re-reviewed exact head `46eb7df096dffec08dc706bb9522a6caa108eebf` with mandatory two-pass review and strict thermo.

Medium: [strict-root validation still does not validate the schema actually emitted by JSON serialization](https://github.com/openai/openai-node/blob/46eb7df096dffec08dc706bb9522a6caa108eebf/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L13-L24). An override with an own-enumerable `type: object` plus `toJSON() { return { type: string } }` passes both descriptor checks, but the request serializes a forbidden string root. Conversely, an enumerable `$ref: undefined` is rejected although JSON omits it. Validate a plain request-equivalent root or reject serialization hooks/exotic accessors, determine keyword presence from emitted data, and add regressions asserting the actual `JSON.stringify` result for both cases.

The prior prototype/non-enumerable keyword issue and reference-engine concerns are resolved; strict thermo and exact CI are clean.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
One Medium issue remains at exact `b1c3bc0558e30cd09c5da0fbf161e82c20bf594a`.

`validateStrictRootSchema` still assumes that any object carrying an own enumerable `type: "object"` serializes as a JSON object. Arrays and boxed primitives use special JSON serialization: `Object.assign([], { type: "object" })` emits `[]`, while boxed numbers/strings/booleans emit primitives. Both strict-root checks therefore accept values that still put a forbidden non-object root on the wire.

Please define one explicit JSON-object-carrier boundary before the descriptor checks (surgically rejecting arrays and boxed primitives is enough) and add focused override regressions asserting actual `JSON.stringify` results for an array and representative boxed primitive. Avoid a full stringify/parse normalization because that would invoke arbitrary hooks/getters and change identity beyond this fix.

The prior `toJSON` and omitted-`$ref` issues are fixed. Strict thermo is otherwise clean (`zodToJsonSchema.ts` 128→186; focused test 404 lines), and all exact-head CI is green.

Anchor: https://github.com/openai/openai-node/blob/b1c3bc0558e30cd09c5da0fbf161e82c20bf594a/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L21-L53

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 5]
```diff
@@ -5,6 +5,59 @@ import { parseDef } from './parseDef';
...
+    break;
+  }
+}
+
+function validateStrictRootSchema(schema: object, openaiStrictMode: boolean | undefined): void {
+  if (!openaiStrictMode) {
+    return;
+  }
+
+  const descriptors = Object.getOwnPropertyDescriptors(schema);
```
[the remark this answers is not in the document]
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in 659779757886505248a3799d4738616202f3cb58. Strict roots now reject real arrays with Array.isArray and boxed String/Number/Boolean values using intrinsic valueOf brand checks without invoking user getters or coercion. Null-prototype object roots remain supported. Added actual JSON.stringify regressions for all four invalid shapes plus a null-prototype control; the focused suite passes 112 tests, and both lint and tsc --noEmit pass.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
One Low edge case remains at exact `659779757886505248a3799d4738616202f3cb58`.

The new JSON-carrier guard rejects arrays plus boxed strings/numbers/booleans, but omits boxed BigInt. `Object.assign(Object(1n), { type: "object" })` passes both strict-root validations, then `JSON.stringify` throws `TypeError: Do not know how to serialize a BigInt`. This is limited to an exotic direct `_vendor` override, so I am treating it as Low, but it still contradicts the guard contract.

Please add `BigInt.prototype.valueOf` to the targeted intrinsic checks and a focused override regression asserting the converter raises `Root schema must serialize to a JSON object` before request serialization. Do not broaden this to every boxed primitive: boxed Symbol emits an object once `type` is assigned and should remain accepted.

The prior array/boxed primitive issue is otherwise fixed. Strict thermo is clean (`zodToJsonSchema.ts` 204 lines; focused test 433 lines), and all exact-head CI is green.

Anchor: https://github.com/openai/openai-node/blob/659779757886505248a3799d4738616202f3cb58/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L26

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 6]
```diff
@@ -5,6 +5,77 @@ import { parseDef } from './parseDef';
...
+    break;
+  }
+}
+
+function validateStrictRootShape(schema: object): void {
+  if (Array.isArray(schema)) {
+    throw new TypeError('Root schema must serialize to a JSON object');
+  }
+
+  for (const unbox of [String.prototype.valueOf, Number.prototype.valueOf, Boolean.prototype.valueOf]) {
```
[the remark this answers is not in the document]
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in 2e732086851e88730d9cee70a879695bfde6f905. Strict root validation now conditionally includes BigInt.prototype.valueOf in its getter-safe intrinsic brand checks, so boxed BigInt schemas fail before JSON serialization while runtimes without BigInt remain supported. Added a focused regression proving JSON.stringify throws for the boxed value and strict conversion rejects it without BigInt literal syntax. All 113 focused tests, repository lint, and tsc --noEmit pass.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
The boxed-BigInt gap is fixed on exact `2e732086851e88730d9cee70a879695bfde6f905`, but this needs one structural correction before another carrier-specific patch.

**[Medium, strict thermo] Define one ordinary JSON-schema record boundary.** `zodToJsonSchema.ts` has grown 128→211 lines and the focused test to 449 lines by approximating JSON serialization piecemeal: prototype walking, arrays, `toJSON`, and four primitive unboxers. Two remaining holes show the catalog is still open-ended:

- **[Low]** A callable root such as `Object.assign(() => {}, { type: "object" })` passes both checks, but `JSON.stringify` returns `undefined`.
- **[Low]** An extensible Proxy whose `getPrototypeOf` trap returns itself serializes as a valid object, but the prototype walk never terminates.

Please stop adding isolated coercion cases and establish one explicit accepted carrier invariant—preferably a plain/ordinary JSON-schema record at the strict typed converter/helper boundary—then reject exotic carriers uniformly. This should remove the unboxer matrix and arbitrary prototype traversal while preserving the focused public-helper/reference regressions. If the callable fix is made locally during that reframing, reject `typeof schema === "function"`; if any prototype walk remains, detect repeats.

All exact-head CI is green and no file crosses 1k, but the repeated special-case growth is the blocker under the repository architecture guidance.

Anchors: https://github.com/openai/openai-node/blob/2e732086851e88730d9cee70a879695bfde6f905/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L8-L84

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 7]
```diff
@@ -5,6 +5,84 @@ import { parseDef } from './parseDef';
 import { getRefs } from './Refs';
 import { zodDef, isEmptyObj } from './util';
 
+function validateStrictRootSerializationHook(schema: object): void {
+  for (let target: object | null = schema; target !== null; target = Object.getPrototypeOf(target)) {
```
[the remark this answers is not in the document]
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in 62b93b2a4a8adeb5edfe2fa252a73546bc15f65b. Serialization-hook inspection now tracks visited prototype objects in an ES2015-compatible WeakSet and rejects repeated Proxy-reported prototypes deterministically without invoking getters. The regression confirms the cyclic Proxy otherwise serializes to an ordinary JSON object and strict validation terminates with a clear error; all 115 focused tests, repository lint, and tsc --noEmit pass.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 8]
```diff
@@ -5,6 +5,84 @@ import { parseDef } from './parseDef';
...
+    }
+    if (!('value' in serializationHook) || typeof serializationHook.value === 'function') {
+      throw new Error("Root schema cannot contain a callable or accessor-backed 'toJSON' property");
+    }
+    break;
+  }
+}
+
+function validateStrictRootShape(schema: object): void {
+  if (Array.isArray(schema)) {
```
[the remark this answers is not in the document]
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in 62b93b2a4a8adeb5edfe2fa252a73546bc15f65b. Strict root validation now rejects callable overrides before descriptor or serialization-hook processing. The new regression confirms a function with enumerable type: 'object' serializes as undefined and is synchronously rejected; all 115 focused tests, repository lint, and tsc --noEmit pass.

File: src/_vendor/zod-to-json-schema/zodToJsonSchema.ts [hunk 9]
```diff
@@ -5,6 +5,90 @@ import { parseDef } from './parseDef';
...
+    const descriptor = descriptors[keyword];
+    if (descriptor?.enumerable && !('value' in descriptor)) {
+      throw new Error(`Root schema validation keyword '${keyword}' must be a data property`);
+    }
+  }
+
+  validateStrictRootSerializationHook(schema);
+
+  const type = descriptors['type']?.enumerable ? descriptors['type'].value : undefined;
+  const nullable = descriptors['nullable']?.enumerable && descriptors['nullable'].value === true;
```
[the remark this answers is not in the document]
Comment (HAYDEN-OAI) [reply src/_vendor/zod-to-json-schema/zodToJsonSchema.ts]: Fixed in 8a545a1b9ed6f5f3a5331a90b57e0c6bc16b2cde. Strict validation now reads own enumerable data-only type, nullable, and $ref keywords through Reflect.get, matching JSON serialization's Proxy Get semantics while continuing to reject accessor descriptors without invoking their getters and to ignore invisible or omitted values. Added a compact four-case serialization regression covering mismatched type, nullable, emitted $ref, and omitted $ref. All 119 focused tests, full repository lint, and tsc --noEmit pass.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
The latest Proxy keyword reads help, but exact `8a545a1b9ed6f5f3a5331a90b57e0c6bc16b2cde` still needs the same Medium boundary redesign rather than more serialization cases.

`zodToJsonSchema.ts` has grown 128→212 lines and the focused suite to 512 lines around a prototype walker, cycle state, array/function cases, four primitive-brand probes, descriptors/accessors, `toJSON`, omitted values, and Proxy reads. This remains an open-ended approximation of ECMAScript serialization rather than one typed root invariant. Concrete holes remain: boxed `nullable: Object(true)` passes but serializes as true; a Proxy can synthesize callable `toJSON` from `get` while descriptors show none; and a fresh Proxy at every `getPrototypeOf` hop never repeats in the WeakSet.

Please define one supported canonical/owned JSON-schema record boundary at the strict converter/helper edge, snapshot/own the root once, validate and return that same stable record, reject unsupported carriers uniformly, and delete the prototype/unboxer catalog. Preserve the canonical `parseDef(..., forceResolution)` path and focused public-helper/registered/lazy/recursive/naming/non-strict regressions. The three examples above are evidence for this single structural fix, not requests for three more local patches.

No file crosses 1k and no suppression/mock issue was found. All changed focused tests pass; the sole Node 22 failure is the unrelated generated Skills Versions APIConnectionError (Node 24/26 and every other applicable check pass).

Anchor: https://github.com/openai/openai-node/blob/8a545a1b9ed6f5f3a5331a90b57e0c6bc16b2cde/src/_vendor/zod-to-json-schema/zodToJsonSchema.ts#L8-L85

**Review by jbeckwith-oai — APPROVED** [review]
Re-reviewed exact head `ef52f13308394db93d8f64fd6af40e9a5705f494` and approved. The prior architecture blocker is resolved: strict conversion now owns one stable plain-record snapshot, validates and returns that same record, and removes the open-ended prototype/unboxer/serialization catalog. Canonical forced parsing, registered/lazy/recursive roots, frozen definitions, naming, public helpers, and non-strict behavior remain sound. Strict thermo is clean at 194 production lines plus a focused 331-line suite, with no suppressions or problematic mocks. The sole Node 22 generated Skills/Versions connection failure is the unrelated mock-server early-404 flake; changed tests, Node 24/26, build, lint, ecosystem, performance, detect-breaking, and CodeQL are green.
